// ionic-image-loader/esm2015/lib/ionic-image-loader.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, Renderer2 } from '@angular/core';
import { ImageLoaderConfigService } from './services/image-loader-config.service';
import { ImageLoaderService } from './services/image-loader.service';
/** @type {?} */
const propMap = {
    display: 'display',
    height: 'height',
    width: 'width',
    backgroundSize: 'background-size',
    backgroundRepeat: 'background-repeat',
};
/**
 * @record
 */
export function ImageAttribute() { }
if (false) {
    /** @type {?} */
    ImageAttribute.prototype.element;
    /** @type {?} */
    ImageAttribute.prototype.value;
}
export class IonicImageLoaderComponent {
    /**
     * @param {?} _element
     * @param {?} renderer
     * @param {?} imageLoader
     * @param {?} config
     */
    constructor(_element, renderer, imageLoader, config) {
        this._element = _element;
        this.renderer = renderer;
        this.imageLoader = imageLoader;
        this.config = config;
        /**
         * Fallback URL to load when the image url fails to load or does not exist.
         */
        this.fallbackUrl = this.config.fallbackUrl;
        /**
         * Whether to show a spinner while the image loads
         */
        this.spinner = this.config.spinnerEnabled;
        /**
         * Whether to show the fallback image instead of a spinner while the image loads
         */
        this.fallbackAsPlaceholder = this.config.fallbackAsPlaceholder;
        /**
         * Attributes to pass through to img tag if _useImg == true
         */
        this.imgAttributes = [];
        /**
         * Enable/Disable caching
         */
        this.cache = true;
        /**
         * Width of the image. This will be ignored if using useImg.
         */
        this.width = this.config.width;
        /**
         * Height of the image. This will be ignored if using useImg.
         */
        this.height = this.config.height;
        /**
         * Display type of the image. This will be ignored if using useImg.
         */
        this.display = this.config.display;
        /**
         * Background size. This will be ignored if using useImg.
         */
        this.backgroundSize = this.config.backgroundSize;
        /**
         * Background repeat. This will be ignored if using useImg.
         */
        this.backgroundRepeat = this.config.backgroundRepeat;
        /**
         * Name of the spinner
         */
        this.spinnerName = this.config.spinnerName;
        /**
         * Color of the spinner
         */
        this.spinnerColor = this.config.spinnerColor;
        /**
         * Notify on image load..
         */
        this.load = new EventEmitter();
        /**
         * Indicates if the image is still loading
         */
        this.isLoading = true;
        this._useImg = this.config.useImg;
    }
    /**
     * Use <img> tag
     * @param {?} val
     * @return {?}
     */
    set useImg(val) {
        this._useImg = val !== false;
    }
    /**
     * Convenience attribute to disable caching
     * @param {?} val
     * @return {?}
     */
    set noCache(val) {
        this.cache = val !== false;
    }
    /**
     * @return {?}
     */
    get src() {
        return this._src;
    }
    /**
     * The URL of the image to load.
     * @param {?} imageUrl
     * @return {?}
     */
    set src(imageUrl) {
        this._src = this.processImageUrl(imageUrl);
        this.updateImage(this._src);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (this.fallbackAsPlaceholder && this.fallbackUrl) {
            this.setImage(this.fallbackUrl, false);
        }
        if (!this.src) {
            // image url was not passed
            // this can happen when [src] is set to a variable that turned out to be undefined
            // one example could be a list of users with their profile pictures
            // in this case, it would be useful to use the fallback image instead
            // if fallbackUrl was used as placeholder we do not need to set it again
            if (!this.fallbackAsPlaceholder && this.fallbackUrl) {
                // we're not going to cache the fallback image since it should be locally saved
                this.setImage(this.fallbackUrl);
            }
            else {
                this.isLoading = false;
            }
        }
    }
    /**
     * @private
     * @param {?} imageUrl
     * @return {?}
     */
    updateImage(imageUrl) {
        this.imageLoader
            .getImagePath(imageUrl)
            .then((url) => this.setImage(url))
            .catch((error) => this.setImage(this.fallbackUrl || imageUrl));
    }
    /**
     * Gets the image URL to be loaded and disables caching if necessary
     * @private
     * @param {?} imageUrl
     * @return {?}
     */
    processImageUrl(imageUrl) {
        if (this.cache === false) {
            // need to disable caching
            if (imageUrl.indexOf('?') < 0) {
                // add ? if doesn't exists
                imageUrl += '?';
            }
            else {
                imageUrl += '&';
            }
            // append timestamp at the end to make URL unique
            imageUrl += 'cache_buster=' + Date.now();
        }
        return imageUrl;
    }
    /**
     * Set the image to be displayed
     * @private
     * @param {?} imageUrl image src
     * @param {?=} stopLoading set to true to mark the image as loaded
     * @return {?}
     */
    setImage(imageUrl, stopLoading = true) {
        this.isLoading = !stopLoading;
        if (this._useImg) {
            // Using <img> tag
            if (!this.element) {
                // create img element if we dont have one
                this.element = this.renderer.createElement('img');
                this.renderer.appendChild(this._element.nativeElement, this.element);
            }
            // set it's src
            this.renderer.setAttribute(this.element, 'src', imageUrl);
            // if imgAttributes are defined, add them to our img element
            this.imgAttributes.forEach((attribute) => {
                this.renderer.setAttribute(this.element, attribute.element, attribute.value);
            });
            if (this.fallbackUrl && !this.imageLoader.nativeAvailable) {
                this.renderer.listen(this.element, 'error', () => this.renderer.setAttribute(this.element, 'src', this.fallbackUrl));
            }
        }
        else {
            // Not using <img> tag
            this.element = this._element.nativeElement;
            for (const prop in propMap) {
                if (this[prop]) {
                    this.renderer.setStyle(this.element, propMap[prop], this[prop]);
                }
            }
            this.renderer.setStyle(this.element, 'background-image', `url("${imageUrl || this.fallbackUrl}")`);
        }
        if (stopLoading) {
            this.load.emit(this);
        }
    }
}
IonicImageLoaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'img-loader',
                template: `
    <ion-spinner
        *ngIf="spinner && isLoading && !fallbackAsPlaceholder"
        [name]="spinnerName"
        [color]="spinnerColor"
    ></ion-spinner>
    <ng-content></ng-content>
  `,
                styles: ['ion-spinner { float: none; margin-left: auto; margin-right: auto; display: block; }']
            }] }
];
/** @nocollapse */
IonicImageLoaderComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: ImageLoaderService },
    { type: ImageLoaderConfigService }
];
IonicImageLoaderComponent.propDecorators = {
    fallbackUrl: [{ type: Input }],
    spinner: [{ type: Input }],
    fallbackAsPlaceholder: [{ type: Input }],
    imgAttributes: [{ type: Input }],
    cache: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    display: [{ type: Input }],
    backgroundSize: [{ type: Input }],
    backgroundRepeat: [{ type: Input }],
    spinnerName: [{ type: Input }],
    spinnerColor: [{ type: Input }],
    load: [{ type: Output }],
    useImg: [{ type: Input }],
    noCache: [{ type: Input }],
    src: [{ type: Input }]
};
if (false) {
    /**
     * Fallback URL to load when the image url fails to load or does not exist.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.fallbackUrl;
    /**
     * Whether to show a spinner while the image loads
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.spinner;
    /**
     * Whether to show the fallback image instead of a spinner while the image loads
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.fallbackAsPlaceholder;
    /**
     * Attributes to pass through to img tag if _useImg == true
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.imgAttributes;
    /**
     * Enable/Disable caching
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.cache;
    /**
     * Width of the image. This will be ignored if using useImg.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.width;
    /**
     * Height of the image. This will be ignored if using useImg.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.height;
    /**
     * Display type of the image. This will be ignored if using useImg.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.display;
    /**
     * Background size. This will be ignored if using useImg.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.backgroundSize;
    /**
     * Background repeat. This will be ignored if using useImg.
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.backgroundRepeat;
    /**
     * Name of the spinner
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.spinnerName;
    /**
     * Color of the spinner
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.spinnerColor;
    /**
     * Notify on image load..
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.load;
    /**
     * Indicates if the image is still loading
     * @type {?}
     */
    IonicImageLoaderComponent.prototype.isLoading;
    /** @type {?} */
    IonicImageLoaderComponent.prototype.element;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype._useImg;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype._src;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype._element;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype.imageLoader;
    /**
     * @type {?}
     * @private
     */
    IonicImageLoaderComponent.prototype.config;
}
