// ionic-image-loader/esm5/lib/ionic-image-loader.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { File } from '@ionic-native/file/ngx';
import { IonicModule } from '@ionic/angular';
import { IonicImageLoaderComponent } from './ionic-image-loader.component';
import { ImageLoaderConfigService } from './services/image-loader-config.service';
import { ImageLoaderService } from './services/image-loader.service';
var IonicImageLoader = /** @class */ (function () {
    function IonicImageLoader() {
    }
    /**
     * @return {?}
     */
    IonicImageLoader.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: IonicImageLoader,
            providers: [
                ImageLoaderConfigService,
                ImageLoaderService,
                File,
            ],
        };
    };
    IonicImageLoader.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        IonicModule,
                        HttpClientModule,
                        CommonModule,
                    ],
                    declarations: [IonicImageLoaderComponent],
                    exports: [IonicImageLoaderComponent],
                },] }
    ];
    return IonicImageLoader;
}());
export { IonicImageLoader };
