// ionic-image-loader/esm5/lib/services/image-loader-config.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var ImageLoaderConfigService = /** @class */ (function () {
    function ImageLoaderConfigService() {
        this.debugMode = false;
        this.spinnerEnabled = true;
        this.fallbackAsPlaceholder = false;
        this.backgroundSize = 'contain';
        this.backgroundRepeat = 'no-repeat';
        this.display = 'block';
        this.width = '100%';
        this.height = '100%';
        this.useImg = false;
        this.concurrency = 5;
        this.maxCacheSize = -1;
        this.maxCacheAge = -1;
        this.imageReturnType = 'uri';
        // Must be default 'true' for the new WebView to show images
        this.fileNameCachedWithExtension = true;
        this.fallbackFileNameCachedExtension = '.jpg';
        this.cacheDirectoryType = 'cache';
        this._cacheDirectoryName = 'image-loader-cache';
    }
    Object.defineProperty(ImageLoaderConfigService.prototype, "cacheDirectoryName", {
        get: /**
         * @return {?}
         */
        function () {
            return this._cacheDirectoryName;
        },
        set: /**
         * @param {?} name
         * @return {?}
         */
        function (name) {
            name.replace(/\W/g, '');
            this._cacheDirectoryName = name;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Enables debug mode to receive console logs, errors, warnings
     */
    /**
     * Enables debug mode to receive console logs, errors, warnings
     * @return {?}
     */
    ImageLoaderConfigService.prototype.enableDebugMode = /**
     * Enables debug mode to receive console logs, errors, warnings
     * @return {?}
     */
    function () {
        this.debugMode = true;
    };
    /**
     * Enable/Disable the spinner by default. Defaults to true.
     * @param enable set to true to enable
     */
    /**
     * Enable/Disable the spinner by default. Defaults to true.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    ImageLoaderConfigService.prototype.enableSpinner = /**
     * Enable/Disable the spinner by default. Defaults to true.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    function (enable) {
        this.spinnerEnabled = enable;
    };
    /**
     * Enable/Disable the fallback image as placeholder instead of the spinner. Defaults to false.
     * @param enable set to true to enable
     */
    /**
     * Enable/Disable the fallback image as placeholder instead of the spinner. Defaults to false.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    ImageLoaderConfigService.prototype.enableFallbackAsPlaceholder = /**
     * Enable/Disable the fallback image as placeholder instead of the spinner. Defaults to false.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    function (enable) {
        this.fallbackAsPlaceholder = enable;
    };
    /**
     * Sets the cache directory name. Defaults to 'image-loader-cache'
     * @param name name of directory
     */
    /**
     * Sets the cache directory name. Defaults to 'image-loader-cache'
     * @param {?} name name of directory
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setCacheDirectoryName = /**
     * Sets the cache directory name. Defaults to 'image-loader-cache'
     * @param {?} name name of directory
     * @return {?}
     */
    function (name) {
        this.cacheDirectoryName = name;
    };
    /**
     * Set default height for images that are not using <img> tag
     * @param height height
     */
    /**
     * Set default height for images that are not using <img> tag
     * @param {?} height height
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setHeight = /**
     * Set default height for images that are not using <img> tag
     * @param {?} height height
     * @return {?}
     */
    function (height) {
        this.height = height;
    };
    /**
     * Set default width for images that are not using <img> tag
     * @param width Width
     */
    /**
     * Set default width for images that are not using <img> tag
     * @param {?} width Width
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setWidth = /**
     * Set default width for images that are not using <img> tag
     * @param {?} width Width
     * @return {?}
     */
    function (width) {
        this.width = width;
    };
    /**
     * Enable display mode for images that are not using <img> tag
     * @param display Display mode
     */
    /**
     * Enable display mode for images that are not using <img> tag
     * @param {?} display Display mode
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setDisplay = /**
     * Enable display mode for images that are not using <img> tag
     * @param {?} display Display mode
     * @return {?}
     */
    function (display) {
        this.display = display;
    };
    /**
     * Use <img> tag by default
     * @param use set to true to use <img> tag by default
     */
    /**
     * Use <img> tag by default
     * @param {?} use set to true to use <img> tag by default
     * @return {?}
     */
    ImageLoaderConfigService.prototype.useImageTag = /**
     * Use <img> tag by default
     * @param {?} use set to true to use <img> tag by default
     * @return {?}
     */
    function (use) {
        this.useImg = use;
    };
    /**
     * Set default background size for images that are not using <img> tag
     * @param backgroundSize Background size
     */
    /**
     * Set default background size for images that are not using <img> tag
     * @param {?} backgroundSize Background size
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setBackgroundSize = /**
     * Set default background size for images that are not using <img> tag
     * @param {?} backgroundSize Background size
     * @return {?}
     */
    function (backgroundSize) {
        this.backgroundSize = backgroundSize;
    };
    /**
     * Set background repeat for images that are not using <img> tag
     * @param backgroundRepeat Background repeat
     */
    /**
     * Set background repeat for images that are not using <img> tag
     * @param {?} backgroundRepeat Background repeat
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setBackgroundRepeat = /**
     * Set background repeat for images that are not using <img> tag
     * @param {?} backgroundRepeat Background repeat
     * @return {?}
     */
    function (backgroundRepeat) {
        this.backgroundRepeat = backgroundRepeat;
    };
    /**
     * Set fallback URL to use when image src is undefined or did not resolve.
     * This image will not be cached. This should ideally be a locally saved image.
     * @param fallbackUrl The remote or local URL of the image
     */
    /**
     * Set fallback URL to use when image src is undefined or did not resolve.
     * This image will not be cached. This should ideally be a locally saved image.
     * @param {?} fallbackUrl The remote or local URL of the image
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setFallbackUrl = /**
     * Set fallback URL to use when image src is undefined or did not resolve.
     * This image will not be cached. This should ideally be a locally saved image.
     * @param {?} fallbackUrl The remote or local URL of the image
     * @return {?}
     */
    function (fallbackUrl) {
        this.fallbackUrl = fallbackUrl;
    };
    /**
     * Set the maximum number of allowed connections at the same time.
     * @param concurrency
     */
    /**
     * Set the maximum number of allowed connections at the same time.
     * @param {?} concurrency
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setConcurrency = /**
     * Set the maximum number of allowed connections at the same time.
     * @param {?} concurrency
     * @return {?}
     */
    function (concurrency) {
        this.concurrency = concurrency;
    };
    /**
     * Sets the maximum allowed cache size
     * @param cacheSize Cache size in bytes
     */
    /**
     * Sets the maximum allowed cache size
     * @param {?} cacheSize Cache size in bytes
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setMaximumCacheSize = /**
     * Sets the maximum allowed cache size
     * @param {?} cacheSize Cache size in bytes
     * @return {?}
     */
    function (cacheSize) {
        this.maxCacheSize = cacheSize;
    };
    /**
     * Sets the maximum allowed cache age
     * @param cacheAge Maximum cache age in milliseconds
     */
    /**
     * Sets the maximum allowed cache age
     * @param {?} cacheAge Maximum cache age in milliseconds
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setMaximumCacheAge = /**
     * Sets the maximum allowed cache age
     * @param {?} cacheAge Maximum cache age in milliseconds
     * @return {?}
     */
    function (cacheAge) {
        this.maxCacheAge = cacheAge;
    };
    /**
     * Set the return type of cached images
     * @param imageReturnType The return type; either 'base64' or 'uri'
     */
    /**
     * Set the return type of cached images
     * @param {?} imageReturnType The return type; either 'base64' or 'uri'
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setImageReturnType = /**
     * Set the return type of cached images
     * @param {?} imageReturnType The return type; either 'base64' or 'uri'
     * @return {?}
     */
    function (imageReturnType) {
        this.imageReturnType = imageReturnType;
    };
    /**
     * Set the default spinner name
     * @param name
     */
    /**
     * Set the default spinner name
     * @param {?} name
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setSpinnerName = /**
     * Set the default spinner name
     * @param {?} name
     * @return {?}
     */
    function (name) {
        this.spinnerName = name;
    };
    /**
     * Set the default spinner color
     * @param color
     */
    /**
     * Set the default spinner color
     * @param {?} color
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setSpinnerColor = /**
     * Set the default spinner color
     * @param {?} color
     * @return {?}
     */
    function (color) {
        this.spinnerColor = color;
    };
    /**
     * Set headers options for the HttpClient transfers.
     * @param headers
     */
    /**
     * Set headers options for the HttpClient transfers.
     * @param {?} headers
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setHttpHeaders = /**
     * Set headers options for the HttpClient transfers.
     * @param {?} headers
     * @return {?}
     */
    function (headers) {
        this.httpHeaders = headers;
    };
    /**
     * Set options for the FileTransfer plugin
     * @param options
     * @deprecated FileTransfer plugin removed.
     */
    /**
     * Set options for the FileTransfer plugin
     * @deprecated FileTransfer plugin removed.
     * @param {?} options
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setFileTransferOptions = /**
     * Set options for the FileTransfer plugin
     * @deprecated FileTransfer plugin removed.
     * @param {?} options
     * @return {?}
     */
    function (options) {
        // do nothing, plugin deprecated.
    };
    /**
     * Enable/Disable the save filename of cached images with extension.  Defaults to false.
     * @param enable set to true to enable
     */
    /**
     * Enable/Disable the save filename of cached images with extension.  Defaults to false.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setFileNameCachedWithExtension = /**
     * Enable/Disable the save filename of cached images with extension.  Defaults to false.
     * @param {?} enable set to true to enable
     * @return {?}
     */
    function (enable) {
        this.fileNameCachedWithExtension = enable;
    };
    /**
     * Set fallback extension filename of cached images.  Defaults to '.jpg'.
     * @param extension fallback extension (e.x .jpg)
     */
    /**
     * Set fallback extension filename of cached images.  Defaults to '.jpg'.
     * @param {?} extension fallback extension (e.x .jpg)
     * @return {?}
     */
    ImageLoaderConfigService.prototype.setFallbackFileNameCachedExtension = /**
     * Set fallback extension filename of cached images.  Defaults to '.jpg'.
     * @param {?} extension fallback extension (e.x .jpg)
     * @return {?}
     */
    function (extension) {
        this.fallbackFileNameCachedExtension = extension;
    };
    ImageLoaderConfigService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                },] }
    ];
    /** @nocollapse */ ImageLoaderConfigService.ngInjectableDef = i0.defineInjectable({ factory: function ImageLoaderConfigService_Factory() { return new ImageLoaderConfigService(); }, token: ImageLoaderConfigService, providedIn: "root" });
    return ImageLoaderConfigService;
}());
export { ImageLoaderConfigService };
if (false) {
    /** @type {?} */
    ImageLoaderConfigService.prototype.debugMode;
    /** @type {?} */
    ImageLoaderConfigService.prototype.spinnerEnabled;
    /** @type {?} */
    ImageLoaderConfigService.prototype.fallbackAsPlaceholder;
    /** @type {?} */
    ImageLoaderConfigService.prototype.backgroundSize;
    /** @type {?} */
    ImageLoaderConfigService.prototype.backgroundRepeat;
    /** @type {?} */
    ImageLoaderConfigService.prototype.display;
    /** @type {?} */
    ImageLoaderConfigService.prototype.width;
    /** @type {?} */
    ImageLoaderConfigService.prototype.height;
    /** @type {?} */
    ImageLoaderConfigService.prototype.useImg;
    /** @type {?} */
    ImageLoaderConfigService.prototype.fallbackUrl;
    /** @type {?} */
    ImageLoaderConfigService.prototype.concurrency;
    /** @type {?} */
    ImageLoaderConfigService.prototype.maxCacheSize;
    /** @type {?} */
    ImageLoaderConfigService.prototype.maxCacheAge;
    /** @type {?} */
    ImageLoaderConfigService.prototype.imageReturnType;
    /** @type {?} */
    ImageLoaderConfigService.prototype.spinnerName;
    /** @type {?} */
    ImageLoaderConfigService.prototype.spinnerColor;
    /** @type {?} */
    ImageLoaderConfigService.prototype.httpHeaders;
    /** @type {?} */
    ImageLoaderConfigService.prototype.fileNameCachedWithExtension;
    /** @type {?} */
    ImageLoaderConfigService.prototype.fallbackFileNameCachedExtension;
    /** @type {?} */
    ImageLoaderConfigService.prototype.cacheDirectoryType;
    /**
     * @type {?}
     * @private
     */
    ImageLoaderConfigService.prototype._cacheDirectoryName;
}
